projects/rulesets-base/src/rules/spectral-rule.ts

```diff
@@ -0,0 +1,137 @@
+import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
+import { IFact, IChange, OpenAPIV3, Result } from '@useoptic/openapi-utilities';
+import { ExternalRuleBase } from './external-rule-base';
+
+type Lifecycle = 'added' | 'addedOrChanged' | 'changed' | 'always';
+type FactTree = Record<string, { fact: IFact | null; children: FactTree }>;
+
+// Spectral results will be returned if they fail
+export interface SpectralResult {
+  path: (string | number)[];
+  code: string | number;
+  message: string;
+}
+
+export interface Spectral {
+  run: (jsonSpec: OpenAPIV3.Document) => Promise<SpectralResult[]>;
+}
+
+function toOpticResult(
+  spectralResult: SpectralResult,
+  lifecycle: Lifecycle,
+  relevantChangeOrFact: IFact | IChange
+): Result {
+  return {
+    condition: spectralResult.code.toString(),
+    passed: false,
+    error: spectralResult.message,
+    isMust: true,
+    isShould: false,
+    where: `${lifecycle} `,
+    change: relevantChangeOrFact,
+  };
+}
+
+function constructFactTree(facts: IFact[]): FactTree {
```

Comment:
Its currently implemented in ruleset-base, but this function will become useful for translating json paths to "openapi nodes"

This takes in facts, which in the future may have different nodes for different api versions.